jscomp/runtime/caml_obj.ml

```diff
@@ -167,32 +163,41 @@ let unsafe_js_compare x y =
 let rec caml_compare (a : Obj.t) (b : Obj.t) : int =
   if a == b then 0 else
   (*front and formoest, we do not compare function values*)
-  if a == (Obj.repr Js.undefined) then -1 else
-  if b == (Obj.repr Js.undefined) then 1 else
-  if a == (Obj.repr Js.null) then -1 else
-  if b == (Obj.repr Js.null) then 1 else
   let a_type = Js.typeof a in 
   let b_type = Js.typeof b in 
-  if a_type = "string" then
-    Pervasives.compare (Obj.magic a : string) (Obj.magic b )
-  else 
-    let is_a_number = a_type = "number" in 
-    let is_b_number = b_type = "number" in 
-    match is_a_number , is_b_number with 
-    | true, true -> 
+  match a_type, b_type with 
+  | "undefined", _ -> - 1
+  | _, "undefined" -> 1 
+  (** [a] is of type string, b can not be None,
+      [a] could be (Some (Some x)) in that case [b] could be [Some None] or [null]
+       so [b] has to be of type string or null *)       
+  | "string", "string" ->   
+      Pervasives.compare (Obj.magic a : string) (Obj.magic b )
+  | "string", _ -> 
```

Comment:
@bobzhang I assume this means that string is the largest type.
In terms of style, I would not jump from the smallest `undefined` to the largest `string` in the next one, but go progressively from small to large, so it's more maintainable over time.
Also, is this assuming lhs and rhs have the same type? I don't think we need to make that assumption. If we don't it's easier to check that all cases are covered.